vm/src/builtins/iter.rs

```diff
@@ -13,6 +14,91 @@ use crate::{
     TypeProtocol,
 };
 
+#[derive(Debug)]
+pub struct PositionIterInternal {
+    pub position: AtomicCell<usize>,
+    /// object or PyNone if exhausted
+    pub obj: PyRwLock<PyObjectRef>,
+}
+
+impl PositionIterInternal {
+    pub fn new(obj: PyObjectRef) -> Self {
+        Self {
+            position: AtomicCell::new(0),
+            obj: PyRwLock::new(obj),
+        }
+    }
+
+    pub fn is_active(&self, vm: &VirtualMachine) -> bool {
+        !vm.is_none(&self.obj.read())
+    }
+
+    pub fn set_state(&self, state: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
+        if self.is_active(vm) {
+            if let Some(i) = state.payload::<PyInt>() {
+                let i = int::try_to_primitive(i.as_bigint(), vm).unwrap_or(0);
+                self.position.store(i);
+                Ok(())
+            } else {
+                Err(vm.new_type_error("an integer is required.".to_owned()))
+            }
+        } else {
+            Ok(())
+        }
+    }
+
+    pub fn reduce(&self, func: PyObjectRef, vm: &VirtualMachine) -> PyObjectRef {
+        if self.is_active(vm) {
+            vm.ctx.new_tuple(vec![
+                func,
+                vm.ctx.new_tuple(vec![self.obj.read().clone()]),
+                vm.ctx.new_int(self.position.load()),
+            ])
+        } else {
+            vm.ctx
+                .new_tuple(vec![func, vm.ctx.new_tuple(vec![vm.ctx.new_list(vec![])])])
```

Comment:
What `__reduce__` returns here is a weird situation. `deque` is very opinionated about always returning a `deque` instance (and the deque iter type accepts nothing else) while others (types edited in this PR) are more open about this (since `iter` doesn't really care what its initialized with).

I think an empty tuple would be the more sensible default for now, even though `list.__reduce__` returns an empty list in CPython.

Reply:
I had check that in CPython, list, set and dict are using empty list for the bulitin iter to construct the empty iterator. As you said, I also think this is not matter, because an exhursted iter should do nothing.